core/src/modules/ops/record.ts

```diff
@@ -290,162 +290,170 @@ export namespace RecordOps {
+              !existingRecordProperty.valueChangedAt ||
+              !existingRecordProperty.rawValue ||
+              rawValue !== existingRecordProperty.rawValue
+                ? now
+                : existingRecordProperty.valueChangedAt,
+            unique: property.unique,
+          } as typeof bulkCreates[number];
+
+          const matchIdx = bulkCreates.findIndex(
+            (bc) =>
+              bc.recordId === record.id &&
+              bc.propertyId === property.id &&
+              bc.position === position
+          );
 
-          if (!property) {
-            if (ignoreMissingProperties) continue;
-            throw new Error(`cannot find a property for id or key \`${key}\``);
+          if (matchIdx >= 0) {
+            bulkCreates.splice(matchIdx, 1, bulkCreate);
+          } else {
+            bulkCreates.push(bulkCreate);
           }
 
-          // add new GrouparooRecord Properties to batch
-          let position = 0;
-          buildQueries: for (const value of h[key]) {
-            if (position > 0 && !property.isArray) {
-              throw new Error(
-                "cannot set multiple record properties for a non-array property"
-              );
-            }
+          position++;
+        }
 
-            const existingRecordProperty = existingRecordProperties.find(
-              (p) =>
-                p.recordId === record.id &&
-                p.propertyId === property.id &&
-                p.position === position
-            );
-            const { rawValue, invalidValue, invalidReason } =
-              await RecordPropertyOps.buildRawValue(
-                value,
-                property.type,
-                existingRecordProperty
-              );
-
-            bulkCreates.push({
-              id: existingRecordProperty
-                ? existingRecordProperty.id
-                : undefined,
-              recordId: record.id,
-              propertyId: property.id,
-              position,
-              rawValue,
-              invalidValue,
-              invalidReason,
-              state: "ready",
-              stateChangedAt: now,
-              confirmedAt: now,
-              startedAt: null,
-              valueChangedAt:
-                !existingRecordProperty ||
-                !existingRecordProperty.valueChangedAt ||
-                !existingRecordProperty.rawValue ||
-                rawValue !== existingRecordProperty.rawValue
-                  ? now
-                  : existingRecordProperty.valueChangedAt,
-              unique: property.unique,
-            });
-
-            position++;
-          }
+        // delete old properties we didn't update
+        existingRecordProperties
+          .filter(
+            (p) =>
+              p.recordId === record.id &&
+              p.propertyId === property.id &&
+              p.position >= position
+          )
+          .map((p) => bulkDeletes.where.id.push(p.id));
+      }
```

Comment:
Slightly unrelated since the previous implementation seems to be doing the same thing: does this mean that when calling `addOrUpdateProperties` with a partial hash, we lose the value of the other properties that were not included in the hash?

I'm seeing the same happen in the old code but I feel like I'm missing something since I don't think the values get cleared when associating imports to existing records🤔